natlib/svcd.c

```diff
@@ -185,3 +186,50 @@ static int svcd_init( lua_State *L )
 
     return 0;
 }
+
+static int svcd_subscribe(lua_State* L)
+{
+    if (lua_gettop(L) != 4) return luaL_error(L, "Expected (targetip, svcid, attrid, on_notify)");
+     
+    lua_getglobal(L, "SVCD");
+    lua_getfield(L, -1, "ivkid");
+     
+    uint16_t ivkid = luaL_checkinteger(L, -1);
+    lua_pop(L, 1);
+
+    uint16_t new_ivkid = ivkid + 1;
+    if (new_ivkid > 65535) {
+        new_ivkid = 0;
+    }
+
+    lua_pushnumber(L, new_ivkid);
+    lua_setfield(L, -1, "ivkid");
+     
+    lua_getfield(L, -1, "oursubs");
+    // put value on the stack
+    lua_pushinteger(L, ivkid);
+    lua_pushvalue(L, 4); // on_notify
+    lua_settable(L, -3);
+    lua_pop(L, 2);
+     
+     
+    lua_pushlightfunction(L, libstorm_net_sendto);
+     
+    lua_getglobal(L, "SVCD");
+    lua_getfield(L, -1, "ncsock");
+    lua_remove(L, -2); // remove SVCD
+     
+     
+    uint8_t msg[7];
+    msg[0] = 1;
+    ((uint16_t*) (msg+1))[0] = luaL_checknumber(L, 2); //svcid
```

Comment:
this is incorrect code for ARM. You are attempting to write a uint16_t to an unaligned section of an unaligned array. At best, the compiler will happen to emit instructions that can support unaligned writes, at worst you will crash the processor with a Hard Fault.

Memory alignment is a subtle problem, so don't feel bad. You have a couple possible solutions:
a) do everything as bytes
b) allocate msg as an array of 8 bytes, with __attribute__((aligned(2))), ignore the first byte, set the second byte to 1 and then write your uint16_t's as you have done - this way they will be aligned
c) declare a packed struct and use that to populate your message. Note you will need to manually pad an extra byte in the front, same as in b